server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -134,6 +135,47 @@ public StoredContext stashContext() {
         };
     }
 
+    /**
+     * Removes the specified transient headers from the current context. The removed transient headers can be
+     * restored by closing the returned {@link StoredContext}. If such a transient header does not exist before
+     * this method is called, it will also not exist after restore (i.e. restoring asserts the existence
+     * and the values for the transient headers, as they existed before invoking this method).
+     */
+    public StoredContext stashTransientContext(Collection<String> transientHeadersToStash) {
+        if (transientHeadersToStash.isEmpty()) {
+            // no-op
+            return () -> {
+            };
+        }
+        final ThreadContextStruct beforeContext = threadLocal.get();
+        final Map<String, Object> stashedTransientHeaders = new HashMap<>();
+        Map<String, Object> newBeforeTransientHeaders = new HashMap<>(beforeContext.transientHeaders);
+        for (String transientHeaderToStash : transientHeadersToStash) {
+            Object stashed = newBeforeTransientHeaders.remove(transientHeaderToStash);
+            if (stashed != null) {
+                stashedTransientHeaders.put(transientHeaderToStash, stashed);
+            }
```

Comment:
> Could we add Objects.requireNonNull(...) in the existing ThreadContext#putTransient method to stop the possibility of null value ever being added as a transient header?

We could. But it feels to me that it diverges from the scope of this PR, so I believe a separate PR is more suitable.
Even so, I'm not sure that the discussions it could spur are worth it.

I have changed the code slightly:
```
        for (String transientHeaderToStash : transientHeadersToStash) {
            if (newBeforeTransientHeaders.containsKey(transientHeaderToStash)) {
                Object stashed = newBeforeTransientHeaders.remove(transientHeaderToStash);
                stashedTransientHeaders.put(transientHeaderToStash, stashed);
            }
        }
```
In order to assuage your original concern that the `stashTransientHeaders` method has slightly different behaviours for null or missing headers.

Reply:
This change is better. Thanks!